gn2/wqflask/templates/oauth2/data-list-mrna.html

```diff
@@ -95,6 +95,8 @@
 	  action="{{search_uri}}"
 	  method="POST">
       <legend>Search: mRNA Assay</legend>
+
+      <input type="hidden" id="bearer_token" name="bearer_token" value="{{bearer_token}}" />
```

Comment:
@jnduli Oof! This induced a panic in me. We do not want to expose the bearer token to the user like this.

Here, the token is sent to the end user, rather than just a session ID, and thus could be stolen and the malicious actor could then pretend to be our user and mess things up.

Look at how I have done most of the other requests: we do not hit the auth-server with JS directly, rather we have:

```example
                                      + <--> GN3
Javascript <--> Browser <--> GN2 <--> +
                                      + <--> gn-auth
```

I would rather you "duplicate" the endpoint on GN2 which will then use the token to pass on the information to gn-auth.

Reply:
Ah, I see. So the ideal workflow would be:

```mermaid
sequenceDiagram
    Actor User
    par login_flow
        User->>GN2: login
        GN2->>GN-auth: login
        GN-auth->>GN2: JWT token
        GN2->>User: SessionId
    and Search Mrna Data to Link
        User->>GN2: SessionId + search_mrna
        GN2->>GN-auth: search_mrna + token
        GN-auth->>GN2: search_result
        GN2->>User: search_resullt
    end
```

why don't we want to have the tokens in the front end? I'd assumed that using tokens ideally supports this where we should be able to call `gn-auth` endpoints directly from any clients since the tokens have expiry times.

For the token getting stolen, I also think the session id also provides such a risk. Have we added any extra measures to ensure this doesn't happen? (e.g. linking the users ip_address to the session id)

Reply:
@jnduli 

> ... So the ideal workflow would be ...

Yes.

> ... why don't we want to have the tokens in the front end? ...

Two reasons:
- The tokens are really short-lived (300 seconds), so there's a very good chance that the token will have expired by the time you get your JS to run and do the requests. The GN2 backend knows to be able to refresh the tokens before doing any requests to the API server. If we expose the token to the browser, we'll need to then implement yet another client in JS to handle the renewal of the tokens.
- Technically, there is nothing wrong with providing a token to the user's browser, but it should be presented in a way that keeps it safe from XSS and/or CSRF attacks. Putting it in the html document makes it so that any JS can easily read the DOM and parse out the token for nefarious purposes. If you must expose the token, consider putting it in [a protected cookie](https://stackoverflow.com/questions/27067251/where-to-store-jwt-in-browser-how-to-protect-against-csrf) that mitigates some of the damage.

> For the token getting stolen, I also think the session id also provides such a risk.

Well, yes and no. The token allows one to directly access the API server. The session ID forces them to go through GN2.

i.e. with the token, you could do something like:

```sh
curl -H 'Authorization: Bearer <your-stolen-token-here>' -H 'application/json' -XPOST https://gn3.endpoint.on.genenetwork.org/ - d ...
```

> Have we added any extra measures to ensure this doesn't happen? (e.g. linking the users ip_address to the session id)

As is the case with these things, I [have left over some space to have this implemented](https://github.com/genenetwork/genenetwork2/blob/0998033d0a7ea26ed96b00a360a334bae6de8c55/gn2/wqflask/oauth2/session.py#L22-L23) but did not actually get around to it. It is not that simple, since a user, could, for example, in one session, log in on their house WIFI, travel to the  office, go to the airport, fly to a different country.

It's still the same computer, and same session, but the IP address will probably flip around like a flag in the wind.

----

TIL: One can build up diagrams in a dot-like syntax on Github: nice!